Improvements (#2)
aerogear/mobile-walkthrough: walkthroughs/1-ionic-showcase/walkthrough.adoc

. Open the `Manage Tasks` page of the showcase app and try create a few tasks. You can open the same URL in a few different tabs and see the tasks synced across all of them in real-time.

[time=15]
== Protect the mobile showcase app by Identity Management service

The showcase app is working, but not secure. Let's use the Identity Management service to protect both the client and the data sync server.

=== Add Identity Management service to the client app

. Go to the link:{route-mdc-server-host}[Mobile Developer Console, window="_blank"].
. Click on the *showcase-app* card to display the configuration screen.
. Click the *Mobile Services* tab.
. Click *Bind to App* for *Identity Management* in the *Unbound Services* section, with the following parameters:
.. Admin User Name: `admin`
.. Admin User Password: `admin`
.. Client Type: `public`
. Close the wizard. Once the binding process is completed, you should the *Identity Management* row is moved to the `Bound Services` section.
. Go back to the `Configuration` tab of the app and you should see the content of `mobile-services.json` is updated. Copy the content and save it somewhere on the computer.

> === Add Identity Management service to the data sync server

Reviewer: This is actually for both client and server. We can split them or have them together (which makes sense)
```suggestion
=== Add Identity Management service to the showcase client and server applications
```

. Go to the `Mobile Services` tab again and expand the *Identity Management* row. Open the `Keycloak Realm URL` in a new tab.
. Login with username `admin` and password `admin`.
. After the console is loaded, we need to create a new client for the showcase data sync server and add some users for the showcase app:
.. Choose `Configure -> clients` in the menu.
.. Click on the `Create` button on the top-right corner, and create a new client with the following parameters:

> ... Client ID: showcase-app-server

Reviewer: I think we might need to have parameter for this in case it will change

Reviewer: Actually do not mind me here.

Reviewer: Actually, do not mind me here.

... Do not change the rest of the fields.
.. Save the client. You should see the *Settings* page of the client.
.. Change the `Access Type` to `bearer-only` and save again.
.. Click on the `Installation` tab, and select `Keycloak OIDC JSON` format. Use the `Download` button to save the configuration file to a directory on the computer.
.. Select `Users` on the left menu, and click on `View all users`. You will see there is only 1 admin user in the realm. Click on `Add user` to create new ones. Pick a username you like and save.
.. Go to `Credentials` tab and set a password for this user. Set `Temporary` option to `OFF`.
. Next, we need to update the showcase sync server app to use the downloaded configuration file:

> .. Go to the link:{openshift-host}/console/project/{walkthrough-namespace}/config-maps[Config Maps page of the OpenShift console], and click on `Create Config Map`.

Reviewer: I think command line path for creating configmap from file will me much simpler TBH. 
`oc create configmap  showcase-server-idm-config --from-file=keycloak.json`

Especially when people will already have oc client pointed to that namespace

Author: yes you are right, but I tried to avoid using CLIs in this walkthrough just incase that they don't have `oc` installed. The idea is that they should be able to do everything within the browser.

Reviewer: ah.. Very valid point!

.. In the creation form, enter the following values:
... Name: showcase-server-idm-config
... Key: keycloak.json
... Value: Use the `Browser` button to load the `keycloak.json` file that was downloaded previously.
.. Once the config map object is created, go to `Applications -> Deployments`. Select the deployment config for `ionic-showcase-server`.
.. Click on the *Configuration* tab, and find the `Volumes` section. Click on the `Add Config Files` at the bottom of the section.
... Choose the `showcase-server-idm-config` config map as the *Source*.
... Set the value for *Mount Path* to be `/tmp/keycloak`.
... Click on `Add`. This will trigger a new deployment.

> .. Go to the *Environment* tab of the deployment config, and add a new environment variable:

Reviewer: We can use `oc set env` here as well.
Going to check how the cmd will look like

Author: the same reason as I commented above.

=== Offline Support

. In order to demo the offline feature, we need to get the device into offline mode first.
.. On a mobile phone this can be done by turning on airplane mode. 
.. In a browser you can use the developer tool to simulate offline mode (For example, in Chrome, you can open it by pressing F12, or using Command+Option+I on macOS. Select the `Network` tab, and check `Offline`).
. Create a new task in the showcase app. The task should be created and the `Offline Changes` button in the footer should contain one change.
. Try make a few more changes by either editing existing tasks, or creating new ones.
. You can review all the changes you have made while offline directly by clicking on `Offline Changes` button.
. Check other instances of the app and you will not see any of the changes you have made because the current one is offline.
. Restore connectivity, and once the client is online again, you should see all `Offline Changes` are synced to the server and appearing in other instances of the showcase app.

=== Conflict resolution

. Make sure there are 2 instances of the app running.
. In both instances of the app:
.. Go offline

> .. Make some changes to the same task

Reviewer: This may no longer trigger conflicts.

```suggestion
.. Make changes on name and description fields for the same tasks on two separate clients
```

> . Bring both of them back online again. You will see a popup showing up in the last instance that is back online warning about conflicts.

Reviewer: ```suggestion
. Bring both of them back online again. You will see a popup showing up in the last instance that is back online warning about conflicts.
. Go offline on both clients
. Make changes on the name field on first client and description on the second one 
. Bring both of them back online again. You should see changes merged on both clients automatically without conflict popup
```

Reviewer: My intention here was to make distinctions between merged and conflicted data
